contracts/FlashMintLib.sol

```diff
@@ -0,0 +1,154 @@
+		(uint256 priceDAIWant, uint256 decimalsDifference, uint256 _requiredDAI) = getPriceDAIWant(want, amountDesired);
+		// to avoid stack too deep	
+		requiredDAI = _requiredDAI;
+		amountWant = amountDesired;
+		// Not enough want in DAI. So we take all we can
+		uint256 _maxFlashLoan = maxLiquidity();
+        if(requiredDAI > _maxFlashLoan) {
+            requiredDAI = _maxFlashLoan.mul(95).div(100); // use only 95% of total liquidity
+            amountWant = requiredDAI.mul(COLLAT_RATIO_DAI).mul(PRICE_DECIMALS).div(priceDAIWant).div(1e18).div(decimalsDifference);
+        }
+	}
+
+	function getPriceDAIWant(address want, uint256 amountDesired) internal returns (uint256 priceDAIWant, uint256 decimalsDifference, uint256 requiredDAI) {
+		uint256 wantDecimals = 10 ** uint256(IERC20Extended(want).decimals());
+		decimalsDifference = DAI_DECIMALS > wantDecimals ? DAI_DECIMALS.div(wantDecimals) : wantDecimals.div(DAI_DECIMALS);
+		if(want == DAI) {
+			requiredDAI = amountDesired.mul(1e18).div(COLLAT_RATIO_DAI);
+			priceDAIWant = PRICE_DECIMALS; // 1:1
+		} else {
+			priceDAIWant = getOraclePrice(DAI).mul(PRICE_DECIMALS).div(getOraclePrice(want));
+			// requiredDAI = desiredWantInDAI / COLLAT_RATIO_DAI
+			// desiredWantInDAI = (desiredWant / priceDAIWant)
+			// NOTE: decimals need adjustment (e.g. BTC: 8 / ETH: 18)
+			requiredDAI = amountDesired.mul(PRICE_DECIMALS).mul(1e18).mul(decimalsDifference).div(priceDAIWant).div(COLLAT_RATIO_DAI);
+		}
+	}
+
+	function getOraclePrice(address token) internal returns (uint256) {
+		string memory symbol;
+		// Symbol for WBTC is BTC in oracle
+		if(token == WBTC) {
+			symbol = "BTC";
+		} else if (token == WETH) {
+			symbol = "ETH";
+		} else {
+            symbol = IERC20Extended(token).symbol();
+        }
+		IUniswapAnchoredView oracle = IUniswapAnchoredView(COMP.oracle());
+		return oracle.price(symbol);
+	}
+
+    event Numbers(string name, uint256 number);
+    event Deficit(bool deficit);
+	function loanLogic(
+		bool deficit,
+        uint256 amountDAI,
+		uint256 amount,
+		CErc20I cToken
+	) 
+        public returns (bytes32)
+    {
+
+        emit Deficit(deficit);
+		// 1. Deposit DAI in Compound as collateral
+		require(CErc20I(CDAI).mint(amountDAI) == 0, "!mint_flash");
+
+		//if in deficit we repay amount and then withdraw
+		if (deficit) {
+			// 2a. if in deficit withdraw amount and repay it
+			require(cToken.redeemUnderlying(amount) == 0, "!redeem_down");
+            (, , uint256 borrowBalance, ) = cToken.getAccountSnapshot(address(this));
+            uint256 amountToRepay = Math.min(borrowBalance, IERC20(cToken.underlying()).balanceOf(address(this)));
+			require(cToken.repayBorrow(amountToRepay) == 0, "!repay_down");
+		} else {
+			// 2b. if levering up borrow and deposit
+			require(cToken.borrow(amount) == 0, "!borrow_up");
+			require(cToken.mint(IERC20(cToken.underlying()).balanceOf(address(this))) == 0, "!mint_up");
```

Comment:
likewise this plus the following redeem is redundant if ctoken is cdai and amount and amountDai is the same